[JENKINS-58176] Create Phase 1 Blog Post (#2349)
jenkins-infra/jenkins.io: content/blog/2019/06/2019-06-25-plugin-management-tool-alpha-release.adoc

---
layout: post
title: "Plugin Management Library and CLI Tool Alpha Release"
tags:
- plugins
- pluginmanagement
- pluginmanager
- platform-sig
- cli
- gsoc
- gsoc2019
author: stopalopa
---
= Plugin Management Library/CLI Tool

_"Everybody is re-inventing the wheel, partially implementing the "details" of plugin management (signed metadata, artifacts checksums, plugins detached from core,...).

> It becomes obvious Jenkins should provide adequate tooling for plugin installation outside jenkins live instance."_ link:https://issues.jenkins-ci.org/browse/JENKINS-53767[JENKINS-53767]

Reviewer: ```suggestion
It becomes obvious Jenkins should provide adequate tooling for plugin installation outside a live Jenkins live instance."_ link:https://issues.jenkins-ci.org/browse/JENKINS-53767[JENKINS-53767]
```

Author: That was a direct quote, but I agree it was not phrased the best. I actually just updated it in Jira to match what you suggested.

> It becomes obvious Jenkins should provide adequate tooling for plugin installation outside a live Jenkins instance."_ link:https://issues.jenkins-ci.org/browse/JENKINS-53767[JENKINS-53767]

Reviewer 2: ```suggestion
It becomes obvious Jenkins should provide adequate tooling for plugin installation outside a live Jenkins instance."_ jira:JENKINS-53767[]
```

should work, but not mandatory

> My link:https://jenkins.io/projects/gsoc/2019/plugin-installation-manager-tool-cli/[Google Summer of Code project]

Reviewer 2: ```suggestion
My link:/projects/gsoc/2019/plugin-installation-manager-tool-cli/[Google Summer of Code project]
```

tries to solve this problem by creating a library that will unify plugin management logic across the different implementations
of Jenkins and providing a CLI tool that will make it easy for users to download plugins and view plugin information before Jenkins even starts.
I'm excited to share that we just released an alpha version that you can check out link:https://github.com/jenkinsci/plugin-installation-manager-tool/releases[here]!

== GSoC Phase 1 Update

While I looked into pulling the link:https://github.com/jenkinsci/jenkins/blob/master/core/src/main/java/hudson/PluginManager.java[Plugin Manager] out of Jenkins core,
this ended up being a challenging first step due to the complexity and number of dependencies.  We instead decided to start by converting the
link:https://github.com/jenkinsci/docker/blob/master/install-plugins.sh[install-plugins.sh bash script] in Jenkins Docker to Java.
There are several issues with the install-plugins.sh script - namely, that it is a bash script and has limited extensibility.
Furthermore, it does not retrieve all of the most-up-to-date update center metadata.

== Alpha Release Details

> Mimicking what was done in the install-plugins.sh script, the new plugin management library takes in a list of plugins, their versions,

Author: Added in 56d7b02.

> Mimicking what was done in the link:https://github.com/jenkinsci/docker/blob/master/install-plugins.sh[install-plugins.sh] script, the new plugin management library takes in a list of plugins, their versions,

Reviewer 2: ```suggestion
Mimicking what was done in the link:https://github.com/jenkinsci/docker/blob/master/install-plugins.sh[install-plugins.sh] script from the official Jenkins Docker image, the new plugin management library takes in a list of plugins, their versions,
```

and/or urls from which to download the plugins, and downloads the requested plugins and their dependencies. The plugins are downloaded from the update center
to a specified directory, and can then be loaded into Jenkins. Currently, the plugins to be downloaded can be specified via a plugins.txt file and/or the -plugins cli option, but we plan to further expand the input formats that can be accepted.
link:https://github.com/jenkinsci/docker#plugin-version-format[Custom version specifiers] for different update centers are also supported.

> image::/images/post-images/gsoc-plugin-management-tool/pluginstxt.png[Example plugins.txt File]

Reviewer: Maybe use carbon sh, for pretty images?
https://github.com/dawnlabs/carbon
e.g.
https://carbon.now.sh/f099cfd39ac60dece0b03bedb4635262

Author: Wow, this is so cool!  Great suggestion.
